### PR TITLE
Extended `mlflow gc` to support removing deleted experiments

### LINKED ISSUE #6617: [FR] Extend `mlflow gc` to support removing deleted experiments
### Willingness to contribute

No. I cannot contribute this feature at this time.

### Proposal Summary

Add an `--experiment-ids` flag to the `mlflow gc` CLI (https://mlflow.org/docs/latest/cli.html#mlflow-gc). For each experiment ID that is specified, `mlflow gc` should remove the experiment and all of its runs. 

Before performing deletion, `mlflow gc` should verify that all of the specified experiments are in the `DELETED` state. `mlflow gc` should support specifying `--experiment-ids` and `--run-ids` at the same time, and `--older-than` should be extended to apply to runs and experiments

### Motivation

> #### What is the use case for this feature?

Enable MLflow admins to permanently remove deleted experiments

> #### Why is this use case valuable to support for MLflow users in general?

The inability to delete experiments means that experiment names cannot be reused and experiment storage space cannot be freed. This FR would address these issues.

> #### Why is this use case valuable to support for your project(s) or organization?

^

> #### Why is it currently difficult to achieve this use case?

MLflow does not currently offer a way to permanently remove deleted experiments

### Details

_No response_

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by shaikmoeed**
## Related Issues/PRs

Resolve #6617 

## What changes are proposed in this pull request?

(Please fill in changes proposed in this fix)

## How is this patch tested?

- [ ] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.
- [x] Not sure how to test

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/cli.py
Comment (harupy) [inline mlflow/cli.py:528]:
```suggestion
    help="Optional comma separated list of experiments to be permanently deleted including "
    "all of its runs. If experiment ids are not specified, data is removed for all "
    "experiments in the `deleted` lifecycle stage. ",
```

File: mlflow/cli.py
Comment (harupy) [inline mlflow/cli.py:528]:
```suggestion
    "experiments in the `deleted` lifecycle stage.",
```

File: mlflow/cli.py
Comment (harupy) [inline mlflow/cli.py:569]: `last_update_time` was added in MLflow 1.29.0. It's `None` for experiments logged before that version. `None <= int_num` results in `TypeError`.

File: mlflow/cli.py
Comment (harupy) [inline mlflow/cli.py:593]: Can we pull these two lines out of the `run_ids` loop?

File: mlflow/cli.py
Comment (harupy) [reply mlflow/cli.py:569]:
```suggestion
        if (experiment.last_update_time or float('inf')) <= time_delta
```

We might need to fallback to `float('inf')`.

Comment (harupy) [reply mlflow/cli.py:569]:
```suggestion
        if experiment.last_update_time is not None or experiment.last_update_time <= time_delta
```

Comment (shaikmoeed) [reply mlflow/cli.py:569]: Can I ignore `if experiment.last_update_time <= time_delta` condition for older versions?

File: mlflow/cli.py
Comment (harupy) [inline mlflow/cli.py:568]: `MlflowClient.search_experiments` is paginated and doesn't fetch all experiments. Can we use `mlflow.search_experiments`?

File: mlflow/cli.py
Comment (harupy) [inline mlflow/cli.py:569]:
```suggestion
        if experiment.last_update_time <= current_time - time_delta
```

Comment (harupy) [reply mlflow/cli.py:569]: I think the expected behavior is:

- If `--older-than` is unspecified, we don't need to look at `last_update_time`. 
- If it's specified, we need to exclude experiments with `last_update_time=None` because we aren't sure when they were created. Some of them might have been created after the specified time threshold.

File: mlflow/cli.py
Comment (shaikmoeed) [reply mlflow/cli.py:568]: For `.search_runs`, which one is recommended `MlflowClient.search_runs` or `mlflow.search_runs # this is returning DataFrame`?

Comment (harupy) [reply mlflow/cli.py:568]: `mlflow.search_runs`. `MlflowClient.search_runs` is also paginated.

File: mlflow/cli.py
Comment (harupy) [inline mlflow/cli.py:576]: I didn't notice we have `backend_store`. Can we use it?

Comment (shaikmoeed) [reply mlflow/cli.py:576]: Sorry, I didn't get it exactly. Do you mean `backend_store` to search experiments?

File: mlflow/cli.py
Comment (harupy) [inline mlflow/cli.py:568]: We have a function to get the current time in milliseconds here: https://github.com/mlflow/mlflow/blob/b26eb133f96fd0cfd4833546490730367ccef46f/mlflow/utils/time_utils.py#L15

File: mlflow/cli.py
Comment (harupy) [reply mlflow/cli.py:576]:
```python
deleted_experiments = []
next_page_token = None
while True:
    experiments = backend_store.search_experiments(
        view_type=ViewType.DELETED_ONLY,
        page_token=next_page_token,
    )
    deleted_experiments.extend(experiments)
    if page_results.next_page_token is None:
        break
    next_page_token = page_results.next_page_token
```

Comment (shaikmoeed) [reply mlflow/cli.py:576]: @harupy Thanks for sharing! I will update. 

Can I use `backend_store` for `search_runs` as well (Any example usage of the same)?

File: mlflow/cli.py
Comment (shaikmoeed) [reply mlflow/cli.py]: Yes, I did.

File: mlflow/cli.py
Comment (harupy) [reply mlflow/cli.py]: Yes, we should do the same for `search_runs`. A helper function like this might be helpful to handle pagination.

```python
def paginate(func, page_token=None, **kwargs):
    page = func(page_token=page_token, **kwargs)
    yield from page
    if page.token:
        yield from paginate(func, page.token, **kwargs)


all_experiments = list(
    paginate(backend_store.search_experiments, view_type=ViewType.DELELTED_ONLY)
)
```

File: mlflow/cli.py
Comment (harupy) [inline mlflow/cli.py:569]:
```suggestion
    time_threshold = get_current_time_millis() - time_delta
```

We can compute the threshold here.

Comment (shaikmoeed) [reply mlflow/cli.py:569]: Okay, I will make the change.

**Comment by harupy** [comment]
@shaikmoeed We have tests for `mlflow gc` in this file:

https://github.com/mlflow/mlflow/blob/b26eb133f96fd0cfd4833546490730367ccef46f/tests/test_cli.py#L259-L270

Can you add tests for `--experiemnt_ids`?

**Comment by shaikmoeed** [comment]
> Can you add tests for `--experiemnt_ids`?

@harupy I am a little new to writing test cases, I can write with your guidance?
What test cases need to cover?

**Comment by harupy** [comment]
> I am a little new to writing test cases, I can write with your guidance?

Got it. Can I push a commit to add one test case? I think that would help understand how to test.

**Comment by shaikmoeed** [comment]
> Got it. Can I push a commit to add one test case? I think that would help you understand how to test.

@harupy  Sure, that helps to get started!

File: mlflow/cli.py
Comment (harupy) [inline mlflow/cli.py:592]: Can we use `backend_store` when fetching runs?

File: mlflow/cli.py
Comment (harupy) [inline mlflow/cli.py:588]:
```suggestion

    if not experiment_ids:
      experiment_ids = []
      next_page_token = None
      time_threshold = get_current_time_millis() - time_delta
      while True:
          page_results = backend_store.search_experiments(
              view_type=ViewType.DELETED_ONLY,
              page_token=next_page_token,
          )
          for experiment in page_results:
              if older_than is None:
                  deleted_experiment_ids.append(experiment.experiment_id)
              elif (
                  experiment.last_update_time is not None
                  and experiment.last_update_time <= time_threshold
              ):
                  experiment_ids.append(experiment.experiment_id)
          if page_results.token is None:
              break
          next_page_token = page_results.token
```

File: mlflow/cli.py
Comment (harupy) [inline mlflow/cli.py:592]: Can we convert `run_ids` to `set` to remove duplicates?

File: mlflow/cli.py
Comment (harupy) [inline mlflow/cli.py:614]: Can we validate the experiment is deleted?

File: tests/test_cli.py
Comment (harupy) [inline tests/test_cli.py:325]:
```suggestion
def test_mlflow_gc_experiments(get_store_details, request):
```

File: mlflow/cli.py
Comment (harupy) [inline mlflow/cli.py:524]: I'm wondering what should happen when both `--experiment-ids` and `--older-than` are specified:

```
mlflow gc --experiment-ids 123 --older-than 1d
```

Should this command behave like "delete experiment 123 if it's older than 1d, otherwise throw"?


@dbczumar Thoughts?

**Comment by harupy** [comment]
@shaikmoeed Good question Mine looks like this:

```json
{
  "python.defaultInterpreterPath": "/home/haru/miniconda3/envs/mlflow-dev-env/bin/python",
  "python.formatting.provider": "black",
  "cSpell.enabled": true,
  "liveServer.settings.port": 5501,
  "prettier.configPath": "mlflow/server/js/.prettierrc.js",
  "python.testing.pytestArgs": ["tests"],
  "python.testing.unittestEnabled": false,
  "python.testing.pytestEnabled": true
}
```

File: mlflow/cli.py
Comment (shaikmoeed) [reply mlflow/cli.py:524]: @harupy I have taken the reference of a similar scenario w.r.t runs, which is throwing an Exception. Do I need to modify both runs and experiments to produce warning messages instead? If so, how warning messages are tested?

Comment (BenWilson2) [reply mlflow/cli.py:524]: Here's an example of using `with pytest.warns()` 
https://github.com/mlflow/mlflow/blob/9f3d3afea71c47657b47ec72def36a04bbb13df3/tests/models/test_model.py#L152-L157

Comment (shaikmoeed) [reply mlflow/cli.py:524]: @BenWilson2 Thanks for sharing an example. What should be the type of warning?

Comment (BenWilson2) [reply mlflow/cli.py:524]: Typically the default UserWarning, when applied to "dear user, I am warning you because of something that you told me to do that I cannot do, sincerely, your computer" is pretty sufficient.

Comment (shaikmoeed) [reply mlflow/cli.py:524]: @BenWilson2  I like the humor. Can I keep the warning message in similar lines, if it's not offensive?

Comment (dbczumar) [reply mlflow/cli.py:524]: @shaikmoeed I think that preserving the same exception behavior used for runs is the best approach here. @BenWilson2 let me know if you have concerns about using an exception rather than a warning. My reasoning here is that the user is asking `mlflow gc` to do something invalid / contradictory: "Delete a particular experiment that is less than X days old and ensure that all deleted experiments are more than X days old".

File: mlflow/cli.py
Comment (dbczumar) [inline mlflow/cli.py:527]:
```suggestion
    help="Optional comma separated list of experiments to be permanently deleted, including "
    "all of their associated runs. If experiment ids are not specified, data is removed for all "
```

File: mlflow/cli.py
Comment (dbczumar) [inline mlflow/cli.py:547]: For backwards compatibility, can we instead skip experiment deletion if the backend doesn't define `_hard_delete_experiment` and print a warning?

File: tests/test_cli.py
Comment (dbczumar) [inline tests/test_cli.py:325]: Can we also add some `search_runs()` validations to this test case ensuring that the experiment's runs have been deleted as well?

**Review by dbczumar — COMMENTED** [review]
@shaikmoeed Thank you for the fantastic PR! I think it's very close to being merge-ready! :D

File: mlflow/cli.py
Comment (harupy) [reply mlflow/cli.py:524]: @BenWilson2 @dbczumar Thanks for the comments. I'd vote for preserving the same exception behavior as Corey suggested.

File: mlflow/cli.py
Comment (shaikmoeed) [reply mlflow/cli.py:547]: @dbczumar @harupy What should be the warning message? Do I need to say like "This cli doesn't have the backend that allows hard-deleting experiments. Update mlflow version >= {current_version}"?

**Comment by shaikmoeed** [comment]
@harupy I have a doubt about pylint settings used in mlflow. It's giving 10/10 for the below code which is not following snake_case for functions.
```
def AbcDef():
    return "some"
```

Is there any reason to ignore this validation check or is it fine to have such names?

**Comment by harupy** [comment]
> Is there any reason to ignore this validation check or is it fine to have such names? If I want to enable this check for my own development work, where can I do it?

No, it's not a good practice to camel-casing for function names. PyLint C103 should prevent this, but it's disabled now.

File: mlflow/cli.py
Comment (harupy) [inline mlflow/cli.py:594]: We've merged https://github.com/mlflow/mlflow/pull/6979. We can use filter experiments by `last_update_time` like this:

```
backend_store.search_experiments(
    view_type=ViewType.DELETED_ONLY,
    filter_string=f"last_update_time < {time_threshold}"
    page_token=next_experiment_page_token,
)
```

**Review by dbczumar — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/6977/files#r991701179 is addressed. Thanks so much, @shaikmoeed !

File: mlflow/cli.py
Comment (harupy) [inline mlflow/cli.py:647]: I think we should perform this validation before deleting experiments:

```python
# Enusre all the experiments are hard-deleted
...


# Then, delete them
for experiment_id in experiment_ids:
    ...
```

File: mlflow/cli.py
Comment (harupy) [inline mlflow/cli.py:548]:
```suggestion
            f" Update mlflow version (current = {version}).",
```

Comment (dbczumar) [reply mlflow/cli.py:548]: @shaikmoeed In this case, updating the MLflow version probably won't help (the maintainer of the third-party backend will need to implement support for hard deletion of experiments). Can we remove the "Update MLflow version" guidance?

File: mlflow/cli.py
Comment (dbczumar) [inline mlflow/cli.py:547]:
```suggestion
            "The specified backend does not allow hard-deleting experiments. Experiments will be skipped."
```

**Review by harupy — APPROVED** [review]
LGTM once the remaining comments are addressed!

File: mlflow/cli.py
Comment (shaikmoeed) [reply mlflow/cli.py:647]: Current version:
```
        for experiment_id in experiment_ids:
            if older_than and experiment_id not in deleted_older_experiment_ids:
                raise MlflowException(
                    f"Experiment {experiment_id} is not older than the required age. "
                    f"Only runs older than {older_than} can be deleted.",
                    error_code=INVALID_PARAMETER_VALUE,
                )
            backend_store._hard_delete_experiment(experiment_id)
```

Suggested version:
```
        for experiment_id in experiment_ids:
            if older_than and experiment_id not in deleted_older_experiment_ids:
                raise MlflowException(
                    f"Experiment {experiment_id} is not older than the required age. "
                    f"Only runs older than {older_than} can be deleted.",
                    error_code=INVALID_PARAMETER_VALUE,
                )
         for experiment_id in experiment_ids:
            backend_store._hard_delete_experiment(experiment_id)
```

@harupy  Is that correct?

Comment (harupy) [reply mlflow/cli.py:647]: We could do something like this:

```python
invalid_experiments = []
for experiment_id in experiment_ids:
    # append to `invalid_experiments` if this experiment can't be deleted

if invalid_experiments:
    # Throw

# Delete experiments
...
```

Comment (shaikmoeed) [reply mlflow/cli.py]: @harupy  I have made the change. But, if we have a long list of `invalid_experiments`, then the error message will be very long. Is that fine?

Comment (harupy) [reply mlflow/cli.py]: I think it's fine.

File: mlflow/cli.py
Comment (harupy) [inline mlflow/cli.py:626]: I think this block needs to be indented.

Comment (shaikmoeed) [reply mlflow/cli.py:626]: I kept this block outside of `if not skip_experiments` to search for runs if experiments are skipped. Do I need to add this block into `if not skip_experiments`?

Comment (harupy) [reply mlflow/cli.py]: Do you think we should delete the runs even when their experiments are NOT going to be deleted?

Comment (shaikmoeed) [reply mlflow/cli.py]: We don't need to delete them. But, we can delete only those runs(with `ViewType.DELETED_ONLY`) which are user mentioned.

**Comment by shaikmoeed** [comment]
@harupy Is there a way to re-run failed check. 

Error:

> requests.exceptions.ConnectionError: HTTPSConnectionPool(host='pypi.python.org', port=443): Max retries exceeded with url: /pypi/scikit-learn/json (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fa445142550>: Failed to establish a new connection: [Errno 101] Network is unreachable'))

File: mlflow/cli.py
Comment (harupy) [inline mlflow/cli.py:608]: `deleted_older_experiment_ids` is only defined when `skip_experiments` is False.

File: mlflow/cli.py
Comment (harupy) [reply mlflow/cli.py]: > We don't need to delete them.

I think the current implementation deletes them.

File: mlflow/cli.py
Comment (shaikmoeed) [reply mlflow/cli.py:626]: @harupy No, I will update the code

Comment (shaikmoeed) [reply mlflow/cli.py]: @harupy Can you check the latest commits?

**Review by harupy — APPROVED** [review]
LGTM!

**Comment by shaikmoeed** [comment]
I hope this was not the longest PR you have ever experienced 🙃

@harupy @dbczumar Thanks for guiding with patience :)

**Pull request merged**
